sw/device/tests/penetrationtests/firmware/hmac_sca.c

```diff
@@ -0,0 +1,222 @@
+  otcrypto_blinded_key_t key = {
+      .config = config,
+      .keyblob_length = sizeof(keyblob),
+      .keyblob = keyblob,
+  };
+
+  // Create input message.
+  otcrypto_const_byte_buf_t input_message = {
+      .len = kMessageLength,
+      .data = msg_buf,
+  };
+
+  // Create tag.
+  otcrypto_word32_buf_t tag = {
+      .len = kTagLengthWord,
+      .data = tag_buf,
+  };
+
+  sca_set_trigger_high();
+  TRY(otcrypto_hmac(&key, input_message, tag));
+  sca_set_trigger_low();
+  return OK_STATUS();
+}
+
+status_t handle_hmac_sca_batch_random(ujson_t *uj) {
+  penetrationtest_hmac_sca_num_it_t uj_it;
+
+  TRY(ujson_deserialize_penetrationtest_hmac_sca_num_it_t(uj, &uj_it));
+
+  uint8_t batch_messages[kNumBatchOpsMax][kMessageLength];
+  uint8_t batch_keys[kNumBatchOpsMax][kKeyLength];
+  uint8_t batch_masks[kNumBatchOpsMax][kKeyLength];
+
+  // Generate random keys and messages.
+  for (size_t it = 0; it < uj_it.num_iterations; it++) {
+    prng_rand_bytes(batch_keys[it], kKeyLength);
+    prng_rand_bytes(batch_masks[it], kKeyLength);
+    prng_rand_bytes(batch_messages[it], kMessageLength);
+  }
+
+  // Invoke HMAC for each data set.
+  uint32_t tag_buf[kTagLengthWord];
+  for (size_t it = 0; it < uj_it.num_iterations; it++) {
+    TRY(trigger_hmac(batch_keys[it], batch_masks[it], batch_messages[it],
+                     tag_buf));
+  }
+
+  // Send the last tag to host via UART.
+  penetrationtest_hmac_sca_tag_t uj_tag;
+  memcpy(uj_tag.tag, tag_buf, kTagLength);
+  RESP_OK(ujson_serialize_penetrationtest_hmac_sca_tag_t, uj, &uj_tag);
+
+  return OK_STATUS();
+}
+
+status_t handle_hmac_sca_batch_fvsr(ujson_t *uj) {
+  penetrationtest_hmac_sca_key_t uj_key;
+  penetrationtest_hmac_sca_num_it_t uj_it;
+
+  TRY(ujson_deserialize_penetrationtest_hmac_sca_key_t(uj, &uj_key));
+  TRY(ujson_deserialize_penetrationtest_hmac_sca_num_it_t(uj, &uj_it));
+
+  uint8_t batch_messages[kNumBatchOpsMax][kMessageLength];
+  uint8_t batch_keys[kNumBatchOpsMax][kKeyLength];
+  uint8_t batch_masks[kNumBatchOpsMax][kKeyLength];
+
+  // First generate all FvsR data sets. When sample_fixed,
+  // the provided key/masks is used and the message is random. When
+  // not sample_fixed, a random key/mask and a random message is
+  // generated.
+  bool sample_fixed = true;
+  for (size_t it = 0; it < uj_it.num_iterations; it++) {
+    if (sample_fixed) {
+      memcpy(batch_keys[it], uj_key.key, kKeyLength);
+      memcpy(batch_masks[it], uj_key.mask, kKeyLength);
+    } else {
+      prng_rand_bytes(batch_keys[it], kKeyLength);
+      prng_rand_bytes(batch_masks[it], kKeyLength);
+    }
+    prng_rand_bytes(batch_messages[it], kMessageLength);
+    sample_fixed = batch_messages[it][0] & 0x1;
+  }
+
+  // Invoke HMAC for each data set.
+  uint32_t tag_buf[kTagLengthWord];
+  for (size_t it = 0; it < uj_it.num_iterations; it++) {
+    TRY(trigger_hmac(batch_keys[it], batch_masks[it], batch_messages[it],
+                     tag_buf));
```

Comment:
IIRC, we do or did chain the results of operations together in batch mode by xoring ciphertexts / digests. As it is now, you'll not notice mismatches happened for the hash operations inside the batch, you only check the last one.

Reply:
Thanks for pointing this out!

The reason why I implemented this was to make sure that the host and the device compute the same random data for the HMAC operation. By sending the last digest, I know when one of the two parties messed up the generation of the random data.

My assumption is that when the host and the device HMAC code get the same random data at its input for each trace, the output should be valid also for each trace.

I know that we do this differently for HMAC (but not for AES). Does it make sense for you to keep it as it is or should I take the XOR of all operations together?

Reply:
The thing is that if the last digest is from the fixed set, it won't tell you anything about the generation of the random data. I would follow the approach taken for KMAC here and XOR them together.

Reply:
Ah, now I notice that this is fvsr_key and that the message is always random. Since you use the same software PRNG for random keys and the always random message, checking the last digest is sufficient actually.

Reply:
No need to make changes then :-)